scripts/functions-deploy-tests/tests.ts

```diff
@@ -0,0 +1,328 @@
+    // All exported functions will have name {hash}-{trigger} e.g. 'abcdefg-v1storage'.
+    await fs.writeFile(
+      path.join(FUNCTIONS_DIR, "index.js"),
+      `export * as ${RUN_ID} from "./fns.js";`
+    );
+  });
+
+  after(async () => {
+    await fs.unlink(path.join(FUNCTIONS_DIR, "index.js"));
+  });
+
+  it("deploys functions with runtime options", async () => {
+    const opts: Opts = {
+      v1Opts: {
+        memory: "128MB",
+        maxInstances: 42,
+        timeoutSeconds: 42,
+      },
+      v2Opts: {
+        memory: "128MiB",
+        maxInstances: 42,
+        timeoutSeconds: 42,
+        // TODO: Re-enable once https://github.com/firebase/firebase-tools/issues/4679 is fixed.
+        // cpu: 2,
+        concurrency: 42,
+      },
+      v1TqOpts: {
+        retryConfig: {
+          maxAttempts: 42,
+          maxRetrySeconds: 42,
+          maxBackoffSeconds: 42,
+          maxDoublings: 42,
+          minBackoffSeconds: 42,
+        },
+        rateLimits: {
+          maxDispatchesPerSecond: 42,
+          maxConcurrentDispatches: 42,
+        },
+      },
+      v2TqOpts: {
+        retryConfig: {
+          maxAttempts: 42,
+          maxRetrySeconds: 42,
+          maxBackoffSeconds: 42,
+          maxDoublings: 42,
+          minBackoffSeconds: 42,
+        },
+        rateLimits: {
+          maxDispatchesPerSecond: 42,
+          maxConcurrentDispatches: 42,
+        },
+      },
+      v1IdpOpts: {
+        blockingOptions: {
+          idToken: true,
+          refreshToken: true,
+          accessToken: false,
+        },
+      },
+      v2IdpOpts: {
+        idToken: true,
+        refreshToken: true,
+        accessToken: true,
+      },
+    };
+    await setOpts(opts);
+
+    const result = await cli.exec(
+      "deploy",
+      FIREBASE_PROJECT,
+      ["--only", "functions", "--non-interactive", "--force"],
+      __dirname,
+      false
+    );
+
+    expect(result.stdout, "deploy result").to.match(/Deploy complete!/);
+
+    const endpoints = await listFns(RUN_ID);
+    expect(Object.keys(endpoints).length, "number of deployed functions").to.equal(FNS_COUNT);
+
+    for (const [id, e] of Object.entries(endpoints)) {
+      if (e.platform === "gcfv1") {
+        expect(e.availableMemoryMb, `${id}.availableMemoryMb`).to.equal(128);
+        expect(e.timeoutSeconds, `${id}.timeoutSeconds`).to.equal(42);
+        expect(e.maxInstances, `${id}maxInstances`).to.equal(42);
+      } else if (e.platform === "gcfv2") {
+        expect(e.availableMemoryMb, `${id}.availableMemoryMb`).to.equal(128);
+        expect(e.timeoutSeconds, `${id}.timeoutSeconds`).to.equal(42);
+        expect(e.maxInstances, `${id}.maxInstances`).to.equal(42);
+        // TODO: Re-enable once https://github.com/firebase/firebase-tools/issues/4679 is fixed.
+        // expect(e.cpu, `${id}.cpu`).to.equal(2);
+        expect(e.concurrency, `${id}.concurrency`).to.equal(42);
+      } else {
+        assertExhaustive(e.platform);
+      }
```

Comment:
What platform would this be, if it's not v1 or v2? `assertExhaustive` isn't a really good name for a test function (that I have to leave the file to go to - and go to a source function at that). Strongly, strongly prefer simply checking for `platform` to be one of the two expected values with `expect` or throwing an error directly here. Importing a one-line function instead of _just_ throwing the error is anti-readability IMO

Reply:
One upside of assertExhaustive is that this gives us a compile time error. If we do end up supporting more platform and changing the type of `platform`, then this code won't compile which is pretty slick if you ask me!

YAGNI? Probably. I think this function is used exclusively in functions portion of the CLI code. Would be interested in hearing if you think we should stop spreading the use of this utility.

Reply:
Regardless - removing this code since other suggestions you made makes this check obsolete.